README.md

```diff
@@ -11,78 +11,114 @@ related-issues: (will contain links to implementation PRs)
 - `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages
 
 # Detailed design
 [design]: #detailed-design
 
-All attributes at the root of nixpkgs (`pkgs.<name>`) whose definition is of the form
-
-```nix
-{
-  <name> = pkgs.callPackage ../some/dir { };
-}
-```
-
-will be become eligible to be transformed as follows:
-- Move `pkgs/some/dir` to `pkgs/auto/<name>`
-- Fix any references to that folder (e.g. in update scripts) 
-- Remove the original definition
-
-These attributes will newly be defined by listing all directories in `pkgs/auto` using `builtins.readDir` and calling `pkgs.callPackage` on all of them, which then gets added to the `pkgs` scope.
-
-Attributes whose definition aren't of the above form won't be changed, so e.g. the following definition in `all-packages.nix` won't be changed:
-```nix
-{
-  syncplay-nogui = syncplay.override { enableGUI = false; };
-}
-```
-
-However, if such definitions can be refactored into the above form they will become eligible for the transformation.
-
-## Backwards compatibility symlinks
-[symlinks]: #backwards-compatibility-symlinks
-
-When moving `pkgs/some/dir/default.nix` to the new `pkgs/unit/<name>/pkg-fun.nix`, a symlink will be created pointing from `pkgs/unit/some/dir/default.nix` to `pkgs/unit/<name>/pkg-fun.nix`. Reasoning:
-- Current community discussions referencing old files from the `master` branch are still valid for some time. While GitHub doesn't provide an easy way to navigate to a symlink, seeing the path to where the file has moved is better than getting an error.
-- It provides an opportunity for code referencing old paths to be updated. While it's not possible to give a deprecation warning with symlinks, users will at least be able to read it in the NixOS release notes. This doesn't occur often in practice.
-
-These symlinks need to be present for at least one NixOS release.
+Make a large part of `pkgs.<name>` definitions in `all-packages.nix` eligible to be moved to `pkgs/unit/<4-letter name>/<name>`.
+The definition in `all-packages.nix` won't be necessary anymore, as all directories in `pkgs/unit/*/*` are automatically added to the `pkgs` set.
+
+The criteria for `pkgs.<name>` becoming eligible are as follows:
+1. <a id="criteria-1"/> Is defined in `pkgs/top-level/all-packages.nix`
+  (necessary so that the overlay containing the automatically discovered packages can be ordered directly before the `all-packages.nix` overlay without changing any behavior)
+2. <a id="criteria-2"/> Is defined to be equal to `pkgs.callPackage <path> { }`
+3. <a id="criteria-3"/> All transitively referenced paths from the default Nix file of `<path>` are under the same directory as the default Nix file and can be moved around together without breaking any references in other Nix files (except the one reference in `pkgs/top-level/all-packages.nix`).
```

Comment:
This is an implementation detail, not an intentional restriction.

Please separate the end goal from what the current tool can do.

Reply:
I guess you're thinking of the tool doing automatic fixing of references in Nix files when moving paths around? At least in all the cases I've seen, that's a violation of API boundaries: Packages shouldn't rely on paths outside of their own "unit" directory. And similarly the other way around. So as far as I can see, it's reasonable that the tool only supports this, and not just for the sake of simplifying the implementation.

Reply:
I think the reader would like to know in a bit more detail what Nixpkgs will look like before diving into the technical details of the process of getting there.

> I guess you're thinking of the tool doing automatic fixing of references in Nix files when moving paths around?

That would be nice, but isn't really the point.

> >  implementation detail, not an intentional restriction

Suppose I want to move `cassandra` into `pkgs/unit`. Its files are referenced by `pkgs.cassandra` and `pkgs.cassandra_3_11`. Am I not allowed to move it, because it violates (2) and (3)?

Reply:
I rewrote the details section a bunch to not mention implementation details anymore.